### PR TITLE
Support expiring access tokens

### PR DESCRIPTION
**Pull request by armstrongnate**
refs: MBL-12228
affects: student, teacher, parent
release note: none

### CODE REVIEW

**Comment by instructure-cx** [comment]
Scan submitted to Checkmarx

**Comment by instructure-cx** [comment]
### Checkmarx scan completed
[Full Scan Details](https://cxmgr.inssec.info/CxWebClient/ViewerMain.aspx?scanid=1005238&projectid=17)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

File: Core/Core/API/API.swift [hunk 1]
```diff
@@ -112,10 +128,43 @@ public struct URLSessionAPI: API {
...
+            let clientSecret = loginSession.clientSecret
+        else {
+            return callback()
+        }
+        let client = APIVerifyClient(authorized: true, base_url: baseURL, client_id: clientID, client_secret: clientSecret)
+        let request = PostLoginOAuthRequest(client: client, refreshToken: refreshToken)
+        makeRequest(request, refreshToken: false) { [weak self] response, _, error in
+            if let response = response, error == nil {
+                let session = loginSession.refresh(accessToken: response.access_token)
+                LoginSession.add(session)
```
Comment (vanwagonet) [inline Core/Core/API/API.swift]: Should it also update the AppEnvironment.currentSession if it matches?

Comment (armstrongnate) [reply Core/Core/API/API.swift]: Yes, good call!

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a release note. If no release note is wanted, use `none`. Example: `release note: Fixed a bug that prevented users from enjoying the app.`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add which apps this change affects. Example: `affects: Teacher, Student` or `affects: none`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a reference to a jira ticket. For example: `refs: MBL-10023`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  One or more files are below the minimum test coverage 50%
  </td>
    </tr>
  </tbody>
</table>



Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 94.35% | 95.02% | -0.66%
Student/Student/CanvasTabBar.swift | 0% | 100% | -100%
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: Core/Core/Extensions/URLResponseExtensions.swift [hunk 2]
```diff
@@ -39,4 +39,8 @@ extension URLResponse {
         }
         return links
     }
+
+    public var unauthorized: Bool {
```
Comment (vanwagonet) [inline Core/Core/Extensions/URLResponseExtensions.swift]: nitpick: `isUnauthorized`?

File: Core/Core/Login/LoginWebPresenter.swift [hunk 3]
```diff
@@ -117,30 +117,24 @@ class LoginWebPresenter {
...
-                        self?.view?.showError(error)
+                DispatchQueue.main.async {
+                    guard let token = response, error == nil else {
+                        self?.view?.showError(error ?? NSError.internalError())
+                        return
                     }
+                    let session = LoginSession(
+                        accessToken: token.access_token,
+                        baseURL: baseURL,
+                        refreshToken: token.refresh_token,
```
Comment (vanwagonet) [inline Core/Core/Login/LoginWebPresenter.swift]: Do we not get an `expires_in`? That could be useful for proactively keeping an active token.

File: Core/Core/Extensions/URLResponseExtensions.swift (hunk 2 above)
Comment (armstrongnate) [reply Core/Core/Extensions/URLResponseExtensions.swift]: Yes!

File: Core/Core/Login/LoginWebPresenter.swift (hunk 3 above)
Comment (armstrongnate) [reply Core/Core/Login/LoginWebPresenter.swift]: On it.

File: CanvasCore/CanvasCore/Extensions/CKIClient+Extensions.swift [hunk 4]
```diff
@@ -36,7 +36,7 @@ extension CKIClient {
         let url = baseURL ?? URL(string: "")!
         var masqueradeID: String? = nil
         if let actAsUserID = actAsUserID, actAsUserID.count > 0 { masqueradeID = actAsUserID }
-        let session = Session(baseURL: url, user: currentUser.sessionUser(), token: accessToken, masqueradeAsUserID: masqueradeID)
+        let session = Session(baseURL: url, user: currentUser.sessionUser(), token: accessToken, refreshToken: refreshToken, clientID: clientID, clientSecret: clientSecret, masqueradeAsUserID: masqueradeID)
```
Comment (armstrongnate) [inline CanvasCore/CanvasCore/Extensions/CKIClient+Extensions.swift]: I just realized that `authSession` is not getting updated when the token refreshes. I think I will have a way to create a `CanvasCore.Session` from a `Core.LoginSession` and replace all uses of `CKIClient.current?.authSession` with `AppEnvironment.shared.currentSession?.authSession`.

**Comment by vanwagonet** [comment]
Ok, I can understand it not working with the French course. It did log me out though after going through another course. I can test that again without going to French first.

**Pull request merged**